extensions/stackdriver/edges/edge_reporter_test.cc

```diff
@@ -0,0 +1,190 @@
+#include "gtest/gtest.h"
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::protobuf::util::MessageDifferencer;
+
+namespace {
+
+#define EXPECT_PROTO_EQUAL(want, got, message)   \
+  std::string diff;                              \
+  MessageDifferencer differ;                     \
+  differ.ReportDifferencesToString(&diff);       \
+  bool equal = differ.Compare(want, got);        \
+  if (!equal) {                                  \
+    std::cerr << message << " " << diff << "\n"; \
+    FAIL();                                      \
+  }                                              \
+  return
+
+std::unique_ptr<ReportTrafficAssertionsRequest> last_request_;
+
+class TestMeshEdgesServiceClient : public MeshEdgesServiceClient {
+ public:
+  typedef std::function<void(std::unique_ptr<ReportTrafficAssertionsRequest>)>
+      TestFn;
+
+  TestMeshEdgesServiceClient(TestFn test_func)
+      : request_callback_(std::move(test_func)){};
+
+  virtual void reportTrafficAssertions(
+      std::unique_ptr<ReportTrafficAssertionsRequest> request) const override {
+    request_callback_(std::move(request));
+  };
+
+ private:
+  TestFn request_callback_;
+};
+
+wasm::common::NodeInfo nodeInfo() {
+  wasm::common::NodeInfo node_info;
+  (*node_info.mutable_platform_metadata())[Common::kGCPProjectKey] =
+      "test_project";
+  (*node_info.mutable_platform_metadata())[Common::kGCPClusterNameKey] =
+      "test_cluster";
+  (*node_info.mutable_platform_metadata())[Common::kGCPClusterLocationKey] =
+      "test_location";
+  node_info.set_namespace_("test_namespace");
+  node_info.set_name("test_pod");
+  node_info.set_workload_name("test_workload");
+  node_info.set_owner("kubernetes://test_owner");
+  return node_info;
+}
+
+wasm::common::NodeInfo peerNodeInfo() {
+  wasm::common::NodeInfo node_info;
+  (*node_info.mutable_platform_metadata())[Common::kGCPProjectKey] =
+      "test_project";
+  (*node_info.mutable_platform_metadata())[Common::kGCPClusterNameKey] =
+      "test_cluster";
+  (*node_info.mutable_platform_metadata())[Common::kGCPClusterLocationKey] =
+      "test_location";
+  node_info.set_namespace_("test_peer_namespace");
+  node_info.set_name("test_peer_pod");
+  node_info.set_workload_name("test_peer_workload");
+  node_info.set_owner("kubernetes://peer_owner");
+  return node_info;
+}
+
+::Wasm::Common::RequestInfo requestInfo() {
+  ::Wasm::Common::RequestInfo request_info;
+  request_info.destination_service_host = "httpbin.org";
+  request_info.request_protocol = "HTTP";
+  return request_info;
+}
+
+std::unique_ptr<ReportTrafficAssertionsRequest> want() {
+  auto request_info = requestInfo();
+  auto peer_node_info = peerNodeInfo();
+  auto node_info = nodeInfo();
+
+  auto req = std::make_unique<ReportTrafficAssertionsRequest>();
+  req->set_parent("projects/test_project");
```

Comment:
This is ok, but you can make it easier by using proto parsing. https://github.com/istio/proxy/blob/master/src/istio/control/tcp/request_handler_impl_test.cc#L89

Reply:
converted where this made sense.